src/v/cluster/scheduling/leader_balancer.cc

```diff
@@ -0,0 +1,519 @@
+    absl::flat_hash_set<raft::group_id> res;
+    for (const auto& e : _muted) {
+        res.insert(e.first);
+    }
+    return res;
+}
+
+std::optional<model::broker_shard>
+leader_balancer::find_leader_shard(const model::ntp& ntp) {
+    /*
+     * look up the broker_shard for ntp's leader. we simply seem to lack
+     * interfaces for making this query concise, but it is rarely needed.
+     */
+    auto leader_node = _leaders.get_leader(ntp);
+    if (!leader_node) {
+        return std::nullopt;
+    }
+
+    auto config_it = _topics.topics_map().find(
+      model::topic_namespace_view(ntp));
+    if (config_it == _topics.topics_map().end()) {
+        return std::nullopt;
+    }
+
+    for (const auto& partition : config_it->second.configuration.assignments) {
+        if (partition.id != ntp.tp.partition) {
+            continue;
+        }
+
+        /*
+         * scan through the replicas until we find the leader node and then
+         * pluck out the shard from that assignment.
+         */
+        for (const auto& replica : partition.replicas) {
+            if (replica.node_id == *leader_node) {
+                return model::broker_shard{replica.node_id, replica.shard};
+            }
+        }
+    }
+
+    return std::nullopt;
+}
+
+/*
+ * builds an index that maps each core in the cluster to the set of replica
+ * groups such that the leader of each mapped replica group is on the given
+ * core. the index is used by a balancing strategy to compute metrics and to
+ * search for leader movements that improve overall balance in the cluster. the
+ * index is computed from controller metadata.
+ */
+leader_balancer::index_type leader_balancer::build_index() const {
+    absl::flat_hash_set<model::broker_shard> cores;
+    index_type index;
+
+    // for each ntp in the cluster
+    for (const auto& topic : _topics.topics_map()) {
+        for (const auto& partition : topic.second.configuration.assignments) {
+            /*
+             * skip balancing for the controller partition, otherwise we might
+             * just constantly move ourselves around.
+             */
+            if (
+              topic.first.ns == model::controller_ntp.ns
+              && topic.first.tp == model::controller_ntp.tp.topic) {
+                continue;
+            }
+            /*
+             * exclude partitions with no known leader from the balancing
+             * strategy. this is done because the balancer does not
+             * currently compute a global schedule of leadership
+             * assignments. and instead evolves the current state one
+             * movement at a time. an alterative approach might be to
+             * compute a schedule and broadcast this, allowing nodes to take
+             * determistic but independent actions.
+             */
+            auto leader_node = _leaders.get_leader(topic.first, partition.id);
+            if (!leader_node) {
```

Comment:
Excluding groups in a transitory leaderless state could make the rebalance a bit over-responsive.  One group temporarily losing its leader shouldn't immediately prompt other groups to adjust leaderships to fill the apparent gap where the leaderless group used to be.

Maybe this needs to carry some state about who the most recent leader was for groups, and time that out after some longer period, something like 5-10mins (order of magnitude of a node reboot).

Reply:
yeh im glad you noticed this. I changed things so that the partition is still represented as a resource in the index (avoiding artificially under counting leaders), but left tracking last-leadership state as a TODO item. instead, I assigned the leader to a random choice from its known replica set, and then exclude the partition from being chosen for rebalancing.